zetta_utils/augmentations/misalign.py

```diff
@@ -0,0 +1,217 @@
+                # so none of the tensor needs to be shifted
+                result = None
+            elif z_chosen < z_diff // 2 and self.mode == "slip":
+                # the chosen Z is falls in the lower cropped out region,
+                # so do nothing for slip
+                result = None
+            else:
+                result = z_chosen - z_diff // 2
+                if result < 0:
+                    assert self.mode == "step"
+                    # "step" when the chosen Z falls in the lowwer cropped
+                    # out region, shift the whole tensor
+                    result = 0
+        else:
+            result = z_chosen
+
+        return result
+
+    def _process_tensor(self, tensor: torch.Tensor, z_size: int, z_chosen: int) -> torch.Tensor:
+        this_z_chosen = self._get_chosen_z_considering_crop(
+            tensor, z_size=z_size, z_chosen=z_chosen
+        )
+
+        assert self.prepared_disp_fraction is not None
+        x_offset = int(tensor.shape[-3] * self.prepared_disp_fraction[0])
+        y_offset = int(tensor.shape[-2] * self.prepared_disp_fraction[1])
+
+        x_size = tensor.shape[-3] - abs(x_offset)
+        y_size = tensor.shape[-2] - abs(y_offset)
+
+        x_start = 0
+        y_start = 0
+        x_start_misal = 0
+        y_start_misal = 0
+
+        if x_offset > 0:
+            x_start += x_offset
+        else:
+            x_start_misal += abs(x_offset)
+
+        if y_offset > 0:
+            y_start += y_offset
+        else:
+            y_start_misal += abs(y_offset)
+
+        if this_z_chosen is not None:
+            if self.mode == "slip":
+                z_misal_slice = slice(this_z_chosen, this_z_chosen + 1)
+            else:
+                z_misal_slice = slice(this_z_chosen, tensor.size(-1))
+
+            tensor[
+                :, x_start : x_start + x_size, y_start : y_start + y_size, z_misal_slice
+            ] = tensor[
+                :,
+                x_start_misal : x_start_misal + x_size,
+                y_start_misal : y_start_misal + y_size,
+                z_misal_slice,
+            ].clone()  # clone necessary bc inplace operation
+        # Crop the data from the pad
+        result = tensor[:, x_start : x_start + x_size, y_start : y_start + y_size, :]
+        return result
+
+    def process_data(self, data: T, mode: Literal["read", "write"]) -> T:
+        if mode != "read":
+            raise NotImplementedError()
+
+        keys_to_apply_final: list[str] | None = None
+        if isinstance(data, torch.Tensor):
+            z_size = data.shape[-1]
+        else:
+            keys_to_apply_final = self._get_keys_to_apply_final(data)
+            sizes = [data[k].shape for k in keys_to_apply_final]
+            if self.disp_in_unit_must_be_divisible_by is None and not all(
+                e[0:2] == sizes[0][0:2] for e in sizes
+            ):
+                raise ValueError(
```

Comment:
I don't fully understand this constraint. The composition of misalignment augmentation is widely used, and misalignment augmentation can always make the volumetric index non-square.